packages/fxa-auth-server/lib/routes/emails.js

```diff
@@ -196,100 +193,94 @@ module.exports = (
           sessionToken.tokenVerified &&
           !email
         ) {
-          return {};
+          return Promise.resolve();
+        }
+
+        const emptyPromise = P.resolve({});
+
+        await customs.check(
+          request,
+          sessionToken.email,
+          'recoveryEmailResendCode'
+        );
+
+        if (!(await setVerifyCode())) {
+          return emptyPromise;
         }
 
+        setVerifyFunction();
+
         const { flowId, flowBeginTime } = await request.app.metricsContext;
 
-        return customs
-          .check(request, sessionToken.email, 'recoveryEmailResendCode')
-          .then(setVerifyCode)
-          .then(setVerifyFunction)
-          .then(() => {
-            if (!sendEmail) {
-              return;
+        const mailerOpts = {
+          code,
+          deviceId: sessionToken.deviceId,
+          flowId,
+          flowBeginTime,
+          service,
+          ip,
+          location: geoData.location,
+          timeZone: geoData.timeZone,
+          timestamp: Date.now(),
+          redirectTo: request.payload.redirectTo,
+          resume: request.payload.resume,
+          acceptLanguage: request.app.acceptLanguage,
+          uaBrowser: sessionToken.uaBrowser,
+          uaBrowserVersion: sessionToken.uaBrowserVersion,
+          uaOS: sessionToken.uaOS,
+          uaOSVersion: sessionToken.uaOSVersion,
+          uaDeviceType: sessionToken.uaDeviceType,
+          uid: sessionToken.uid,
+          style,
+        };
+
+        await verifyFunction(emails, sessionToken, mailerOpts);
+        await request.emitMetricsEvent(`email.${event}.resent`);
+        return emptyPromise;
+
+        // Returns a boolean to indicate whether to send email.
+        async function setVerifyCode() {
```

Comment:
Well done on incremental improvement, this stuff is dense. The myriad ways `code` and `emails` can be set is *really* hard to follow. I tried refactoring this function locally so that it returned the code/emails instead of depending on spooky action at a distance, and failed. I'm wondering whether your improvements can be further simplified, preventing us from shooting ourselves in the foot because of complexity.